cirq/google/engine/engine.py

```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
+            devices when scheduling jobs and gathering calibration metrics.
+
+        Params:
+            project_id: The ID of the Google Cloud project to check, e.g.
+                `my-project-123`
+
+        Returns:
+            A list of dictionaries containing the metadata of each processor.
+        """
+        parent = 'projects/%s' % (project_id)
+        response = self.service.projects().processors().list(
+            parent=parent).execute()
+        return response['processors']
+
+    def get_latest_calibration(self,
+                               processor_name: str) -> Optional['Calibration']:
+        """ Returns metadata about a the latest known calibration run for a
+        processor, or None if there is no calibration available.
+
+        Params:
+            processor_name: A string of the form
+                `projects/project_id/processors/processor_id`.
+
+        Returns:
+            A dictionary containing the metadata.
+        """
+        response = self.service.projects().processors().calibrations().list(
+            parent=processor_name).execute()
+        if (not 'calibrations' in response or
+                len(response['calibrations']) < 1):
+            return None
+        return Calibration(response['calibrations'][0]['data'])
+
+    def get_calibration(self, calibration_name: str) -> 'Calibration':
+        """Retrieve metadata about a specific calibration run.
+
+        Params:
+            calibration_name: A string of the form
+                `<processor name>/calibrations/<ms since epoch>`
+
+        Returns:
+            A dictionary containing the metadata.
+        """
+        response = self.service.projects().processors().calibrations().get(
+            name=calibration_name).execute()
+        return Calibration(response['data']['data'])
+
+
+class Calibration:
+    """A convenience wrapper for calibrations"""
+
+    def __init__(self, calibration: Dict) -> None:
+        self._time = int(calibration['timestampMs'])
+        self._metrics = calibration['metrics']
+
+    def get_timestamp(self) -> int:
+        """ Returns the time of this calibration in milliseconds since
+            the epoch.
+        """
+        return self._time
+
+    def get_metric_names(self) -> List[str]:
+        """ Returns a list of known metrics in this calibration. """
+        return list(set(m['name'] for m in self._metrics))
+
+    def get_metrics_by_name(self, name: str) -> List[Dict]:
+        """ Get a filtered list of metrics matching the provided name.
+
+        Values are grouped into a flat list, grouped by target.
+
+        Params:
+            name: the name of a metric referred to in the calibration. Valid
+            names can be found with the get_metric_names() method.
+
+        Returns:
+            A list of dictionaries containing pairs of targets and values for
+            the requested metric.
+        """
+        return list({
```

Comment:
my preference would be to write this as loops.  when these get multiple lines I find them harder to read.

Reply:
Rewritten to use mutation + loops. WDYT?